flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatisticsTracker.java

```diff
@@ -30,71 +42,99 @@
+  private final StatisticsType statisticsType;
+  private final int switchToSketchThreshold;
+  private final Comparator<StructLike> comparator;
+
   private final Set<Integer> inProgressSubtaskSet;
-  private volatile AggregatedStatistics<D, S> inProgressStatistics;
+  private volatile long inProgressCheckpointId;
+  private volatile StatisticsType coordinatorStatisticsType;
+  private volatile Map<SortKey, Long> coordinatorMapStatistics;
+  private volatile ReservoirItemsUnion<SortKey> coordinatorSketchStatistics;
 
   AggregatedStatisticsTracker(
       String operatorName,
-      TypeSerializer<DataStatistics<D, S>> statisticsSerializer,
-      int parallelism) {
+      int parallelism,
+      Schema schema,
+      SortOrder sortOrder,
+      int downstreamParallelism,
+      StatisticsType statisticsType,
+      int switchToSketchThreshold,
+      @Nullable AggregatedStatistics restoredStatistics) {
     this.operatorName = operatorName;
-    this.statisticsSerializer = statisticsSerializer;
     this.parallelism = parallelism;
+    this.statisticsSerializer =
+        new DataStatisticsSerializer(new SortKeySerializer(schema, sortOrder));
+    this.downstreamParallelism = downstreamParallelism;
+    this.statisticsType = statisticsType;
+    this.switchToSketchThreshold = switchToSketchThreshold;
+
+    this.comparator = SortOrderComparators.forSchema(schema, sortOrder);
     this.inProgressSubtaskSet = Sets.newHashSet();
+    this.coordinatorStatisticsType = StatisticsUtil.collectType(statisticsType, restoredStatistics);
+    this.inProgressCheckpointId = CheckpointStoreUtil.INVALID_CHECKPOINT_ID;
   }
 
-  AggregatedStatistics<D, S> updateAndCheckCompletion(
-      int subtask, DataStatisticsEvent<D, S> event) {
+  AggregatedStatistics updateAndCheckCompletion(int subtask, StatisticsEvent event) {
     long checkpointId = event.checkpointId();
+    LOG.debug(
+        "Handling statistics event from subtask {} of operator {} for checkpoint {}",
+        subtask,
+        operatorName,
+        checkpointId);
 
-    if (inProgressStatistics != null && inProgressStatistics.checkpointId() > checkpointId) {
+    if (inProgressCheckpointId > checkpointId) {
       LOG.info(
-          "Expect data statistics for operator {} checkpoint {}, but receive event from older checkpoint {}. Ignore it.",
+          "Ignore stale statistics event from operator {} subtask {} for older checkpoint {}. Was expecting data statistics from checkpoint {}",
           operatorName,
-          inProgressStatistics.checkpointId(),
-          checkpointId);
+          subtask,
+          checkpointId,
+          inProgressCheckpointId);
       return null;
     }
 
-    AggregatedStatistics<D, S> completedStatistics = null;
-    if (inProgressStatistics != null && inProgressStatistics.checkpointId() < checkpointId) {
+    AggregatedStatistics completedStatistics = null;
+    if (inProgress() && inProgressCheckpointId < checkpointId) {
```

Comment:
I am not entirely comfortable with these checks.
If I understand correctly, we only aggregate the statistics the latest checkpoint. We either emit a partial statistics for the old one, or just throw it away.

Is it so rare of an edge case for the statistics to arrive out of order? Even with unaligned checkpoints?

Reply:
> Is it so rare of an edge case for the statistics to arrive out of order? Even with unaligned checkpoints?

Unaligned checkpoints probably don't really matter here. I definitely see your concern especially if concurrent checkpoints are set. Default is 1.
```
execution.checkpointing.max-concurrent-checkpoints=1
``` 

To make things more generic, I will change the implementation to support concurrent checkpoints by using Map to track aggregation per checkpoint.

Reply:
this has been addressed. we don't need the complex logic of deciding if partial results should be used.